File: chapitre-25/exemple-nonblock-2.py
#! /usr/bin/python
# ------------------------------------------------------------------
# exemple-nonblock-2.py
# Fichier d'exemple du livre "Developpement Systeme sous Linux"
# (C) 2015-2019 - Christophe BLAESS <christophe@blaess.fr>
# https://www.blaess.fr/christophe/
# ------------------------------------------------------------------

from __future__ import print_function
import os
import fcntl
import time

(fd_r, fd_w) = os.pipe()

if os.fork() == 0:
	while True:
		os.write(fd_w, 'c')
		time.sleep(0.7)
else:
	fcntl.fcntl(fd_r, fcntl.F_SETFL, os.O_NONBLOCK)
	while True:
		try:
			r = os.read(fd_r, 1)
			print("Ok")
		except:
			print("Error")
		time.sleep(0.1)

File: chapitre-25/exemple-nonblock-2.c
// ------------------------------------------------------------------
// exemple-nonblock-2.c
// Fichier d'exemple du livre "Developpement Systeme sous Linux"
// (C) 2000-2019 - Christophe BLAESS <christophe@blaess.fr>
// https://www.blaess.fr/christophe/
// ------------------------------------------------------------------

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main (void)
{
	int  pipe_fd[2];
	char c;

	if (pipe(pipe_fd) !=0) {
		perror("pipe");
		exit(EXIT_FAILURE);
	}
	switch (fork()) {
		case -1 :
			perror("fork");
			exit(EXIT_FAILURE);
		case 0 : /* fils : ecriture */
			close(pipe_fd[0]);
			while (1) {
				write(pipe_fd[1], & c, 1);
				usleep(700000);
			}
		default : /* pere : lecture */
			close(pipe_fd[1]);
			fcntl(pipe_fd[0], F_SETFL, O_NONBLOCK);
			while (1) {
				if (read(pipe_fd[0], & c, 1) == 1) 
					printf("Ok\n");
				else
					printf("Non\n");
				usleep(100000);
			}
	}

	return EXIT_SUCCESS;
}

